ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -866,18 +880,112 @@ RaftClientReply waitForReply(RaftClientRequest request, CompletableFuture<RaftCl
 
   @Override
   public RaftClientReply transferLeadership(TransferLeadershipRequest request) throws IOException {
-    //TODO(runzhiwang): implement transfer leadership in server
-    return null;
+    return waitForReply(request, transferLeadershipAsync(request));
+  }
+
+  private CompletableFuture<RaftClientReply> logAndReturnTransferLeadershipFail(
+      TransferLeadershipRequest request, String msg) {
+    LOG.warn(msg);
+    return CompletableFuture.completedFuture(newExceptionReply(request, new StateMachineException(msg)));
+  }
+
+  boolean isSteppingDown() {
+    return finishTransferLeader != null;
+  }
+
+  private Consumer<RaftPeerId> finishTransferLeader;
+
+  public Consumer<RaftPeerId> finishTransferLeader() {
+    return finishTransferLeader;
+  }
+
+  public Consumer<RaftPeerId> setFinishTransferLeader(Consumer<RaftPeerId> consumer) {
+    return finishTransferLeader = consumer;
+  }
+
+  private void timeoutTransferLeadership(
+      TransferLeadershipRequest request, CompletableFuture<RaftClientReply> replyFuture)
+      throws StateMachineException {
+    synchronized (replyFuture) {
```

Comment:
It is better to use `AtomicReference` instead of `synchronized`.

Reply:
Add `synchroized `is to avoid multi-thread compete at the same time, replyFuture also complete in the following lambda code called by ServerState#setLeader. If without synchroized, the replyFuture maybe complete two times at the same time, even though does not cause error, but maybe we should avoid this.
```
      currLeader -> {
        synchronized (replyFuture) {
          if (currLeader == null || replyFuture.isDone()) {
            return;
          }

          if (currLeader.equals(request.getNewLeader())) {
            replyFuture.complete(newSuccessReply(request));
            setFinishTransferLeader(null);
          }
        }
      };
```

Reply:
With the new `AtomicReference<PendingRequest> pending`, we do not need synchronized anymore, thanks a lot.